### DISCUSSION TITLE
Keybinding UI editor

### DISCUSSION THREAD

**Discussion by JosephTLyons** [zed-industries/zed#23955] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
It isn't always super easy to work with the `keymap.json`.  Sometimes I have no issues, other times, I'm totally lost on why something isn't working.  There is [an issue open for validation of config files](https://github.com/zed-industries/feedback/issues/101), but I'm starting to think a keybinding user interface would be a much nicer experience.

**Describe the solution you'd like**
A panel (maybe something that just opens up in a tab?) that has the following setup (description could be skipped):

|Action|Description|Keybinding|
|-|-|-|
|Go To Line|Moves Cursor to Line|`alt-g`|

The only editable area would be the `Keybinding` column.  At a minimum, the user could just type in the keybinding, but having it capture the binding would be icing on the cake.  

Some additional considerations:
- Ability to detect when there's a collision and prevent users from saving it until it is resolved.
    - Currently, if you have keybinding `A` registered to some action, adding a new keybinding `B` for that action` does not de-register keybinding `A`, so you end up with 2 keybindings doing the same action and you can't de-register the original.  Having the UI enforce that all bindings must be unique would solve this.
- Ability to de-register bindings that they'd never use that they just want to get rid of
- Ability to restore default keybindings

**Comment by iamnbutler** [comment]
Have you seen any good implementations of this? Personally I find VSCode's settings ui abysmal, and dread any time I have to use it.

I feel like the closest positive experience I've had is remapping UIs in games.

**Comment by endigma** [comment]
I've never had any issues with VSCode's keybind editor, it's not in the normal settings area. I like the approach taken with binding *keys* to *commands* rather than *commands* to *keys* like `"down"` => `"cursorDown"` and also a million other things.

There is a UI specifically for this in VSCode, I think it works well.

<img width="1892" alt="image" src="https://user-images.githubusercontent.com/34945114/204881027-96d601a4-1551-46b3-ba2d-e336642005a8.png">

**Comment by admsmz** [comment]
Hi everyone. I made a mockup of a keybinding interface that I believe would be an improvement on the status quo. Basically the idea is to use a table rather than a list. Here's a screenshot:

<img width="1592" alt="Screenshot 2023-02-04 at 00 09 48" src="https://user-images.githubusercontent.com/76435414/216728428-0478dd39-99a5-40aa-9dc0-c6bac9a1acd8.png">

Using two dimensions means you can visualise where approximately a keybinding is based on theme, if many similar keybindings are grouped close together. At least, that's how my mind works. Ymmv.

**Comment by JosephTLyons** [comment]
Im wondering if we can ship something smaller in the interim - just a modal that displays the keys being held, allowing for copying and pasting into the clipboard so you can paste it into your `keymap.json` file.

**Comment by samholmes** [comment]
VS Code's solution is good, but what would be great?

What if instead of a big ol' table, we got an interactive Keymap UI? Instead of opening a settings screen and tinkering our keymaps only to return _back to our flow state_ we could interactively add keymaps without breaking flow state; how could this be done? Here's a proposal.

At anytime in the editor, one could activate the Keymap UI with a single keybinding (say it's `cmd+shift+:`, or `cmd+shift+?` or something not taken). This will bring up a palette like the command palette, showing all the key maps for the current context.  From this UI, there is an focused input which a search term for a particular command can be found (exactly like the command palette). Once a command is found, the user can press the `tab` key to enter a keybinding text. Once they've entered the keybinding text, then can either press `tab` again to go to change the context, or press `enter`/`return` to submit the new keybinding to the list. At any time, `esc` key will exit the Keymap UI.

This Keymap UI palette (or sidebar?) could visually contain a sorted list by bindings with all keybinding matching the current context, and then sections with lists for transitive context. The sections are sorted by the context with the highest match to the current context. This will help aid in discovering which keymaps are available under the current context and any transitive matches. In addition to discovering current keymaps, it will help with discovering what actually exists by just getting the editor in the particular desired state. Some tooltips could help aid in explaining the contexts shown in this UI to understand why they're present.

I like the idea of having a Keymap UI aid in discoverability of the editor's defaults as well as searching for commands or contexts all while not breaking out of the current flow state. Perhaps instead of a separate palette, these features could be extended to the command palette (`cmd+shift+p`) and by pressing tab, this Keymaps mode is entered showing the list of keybindings sorted as mentioned.

**Reply by endigma** [reply]
Interesting idea, but I think implementing exclusively this interactive approach would make it significantly harder to manage keys that do multiple things, especially with chording, and also be a large detriment to keybind discoverability because command palettes are inherently indexed by command and not by the keyboard shortcut.

Perhaps this would be a good UX addition to the command palette, for as you say "in flow" keybind changes, but I believe a high level "here is every keyboard shortcut" in the key->command format is necessary as well, for advanced usage, mass rebinding and shortcut discovery.

**Comment by JosephTLyons** [comment]
Zed has had a keymap editor for some time now. Shoutout to @Anthony-Eid, @probably-neb, and co.
